### PR TITLE
Bluetooth: L2CAP: Remove use of k_sem for credits

### LINKED ISSUE #19922: Linear time to give L2CAP credits
We observe an Android master issuing an L2CAP Connection Request, granting us (as peripheral) ~65000 credits. That results in a call to `l2cap_chan_tx_give_credits()` that takes a long time to return and causes connection time out and 100% CPU load for whole duration (>4 seconds for us).

Credits are implemented using a counting semphore, but the BT Host code is currently only able to increment it by 1 -- leading to 65000 calls to `k_sem_give()`. 

Relevant snippet of `l2cap_chan_tx_give_credits()`:
```
	while (credits--) {
		k_sem_give(&chan->tx.credits); 
	}
```

Similar issue exists for `l2cap_chan_rx_give_credits()`.

Potential solutions:

1. Extend the kernel semaphore give+take api to an additional argument, the value to {in,de}crement by. 
2. Extend the kernel semaphore api with a `give_init` taking the initial resource-count as argument. To just be able to initialize the resource count in O(1) time, rather than O(n). Or can we use `k_sem_init()` for this?
3. Go through Host code and replace the credit count by something adhoc (checking a counter, then using a binary semphore), rather than a counting semaphore.

### PR DESCRIPTION
**Pull request by Vudentz**
With the changes that introduced a queue k_sem is only used with
K_NO_WAIT which means it is no longer possible to wait/block for credits
so the usage of k_sem is no longer needed and can be safely replaced
with atomic_t just to count the available credits at a given instant.

Fixes #19922

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

### CODE REVIEW

File: subsys/bluetooth/host/l2cap.c [hunk 1]
```diff
@@ -1256,6 +1252,25 @@ static void l2cap_chan_seg_sent(struct bt_conn *conn, void *user_data)
...
+		k_spin_unlock(&lock, key);
+		return false;
+	}
+
+	target--;
+
+	k_spin_unlock(&lock, key);
+
+	return true;
+}
```
Comment (jhedberg) [inline subsys/bluetooth/host/l2cap.c]: `l2cap.c` doesn't seem the right place for this. Especially when it uses the reserved `atomic_*` namespace. I'd suggest to split this off into its own patch that adds it to `include/sys/atomic.h`

**Review by jhedberg — CHANGES_REQUESTED** [review]
Looks generally fine, except the for wrong placement of the new atomic function

File: subsys/bluetooth/host/l2cap.c (hunk 1 above)
Comment (Vudentz) [reply subsys/bluetooth/host/l2cap.c]: I was a little hesitant to do that right away since it is not always easy to get things in the kernel side, but perhaps on the atomic.h that would be acceptable.

Comment (jhedberg) [reply subsys/bluetooth/host/l2cap.c]: Btw, the implementation doesn't look correct since it's looking at and decrementing the pointer rather than the value behind it

Comment (Vudentz) [reply subsys/bluetooth/host/l2cap.c]: I experiement a little bit with it residing on atomic.h, the problem is that there is no built-in for it so it needs to be inline and add includes for spinlock, so I don't think it is very clean:

https://gist.github.com/Vudentz/a6b7a1cf93d4e984641107293f6ee900

Comment (jhedberg) [reply subsys/bluetooth/host/l2cap.c]: > I experiement a little bit with it residing on atomic.h, the problem is that there is no built-in for it so it needs to be inline and add includes for spinlock, so I don't think it is very clean

I've added @andrewboie and @andyross as reviewers (marked as codeowners for atomic.h) who can hopefully provide some guidance here. Why do you need to have duplicate implementations in the header and c-file? Why can't you just have a prototype in the header and the implementation in the c-file (which could then use the same lock as other functions there)?

Comment (andrewboie) [reply subsys/bluetooth/host/l2cap.c]: @jhedberg so there's three options for atomic ops, selected in a menu in kernel/Kconfig:

- CONFIG_ATOMIC_OPERATIONS_BUILTIN - This uses the compiler's builtin functions for atomic operations. This is the preferred option, but not all compilers have builtins so we have to fallback on some other options.
- CONFIG_ATOMIC_OPERATIONS_CUSTOM - The arch code has custom assembly routines that Zephyr maintains for these functions.
- CONFIG_ATOMIC_OPERATIONS_C - The worst option. These grab a spinlock and then just do the math non-atomically. Because spinlocks are reserved for kernel use, user mode has to make syscalls. Pretty much for CPUs that lack native atomic instructions.

If you look at the #ifdefery in the header you will see how these implementations are managed, the builtin option has these as simple inlines which translate to the GCC atomics.

Comment (andrewboie) [reply subsys/bluetooth/host/l2cap.c]: BTW https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html GCC does not have a builtin with "test and dec" semantics so I urge you guys to consider re-doing your code to use other atomic ops to get what you need.

File: subsys/bluetooth/host/l2cap.c [hunk 2]
```diff
@@ -1256,6 +1252,25 @@ static void l2cap_chan_seg_sent(struct bt_conn *conn, void *user_data)
...
 }
 
+static bool test_and_dec(atomic_t *target)
+{
+	static struct k_spinlock lock;
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&lock);
+
+	if (!(*target)) {
```
Comment (mped-oticon) [inline subsys/bluetooth/host/l2cap.c]: I realize the spin lock is sufficient, but this currently only works because atomic_t is typedef'd to unsigned int.
The abstraction will not break in future if we use atomic_get, atomic_dec (below).

**Review by mped-oticon — APPROVED** [review]
LGTM. Thanks!

File: subsys/bluetooth/host/l2cap.c [hunk 3]
```diff
@@ -1256,6 +1252,25 @@ static void l2cap_chan_seg_sent(struct bt_conn *conn, void *user_data)
 	l2cap_chan_tx_resume(BT_L2CAP_LE_CHAN(chan));
 }
 
+static bool test_and_dec(atomic_t *target)
+{
+	static struct k_spinlock lock;
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&lock);
```
Comment (joerchan) [inline subsys/bluetooth/host/l2cap.c]: could we use atomic_get and atomic_cas and not have the spin-lock?

File: subsys/bluetooth/host/l2cap.c [hunk 4]
```diff
@@ -1594,7 +1607,7 @@ static void l2cap_chan_update_credits(struct bt_l2cap_le_chan *chan,
 	/* Restore enough credits to complete the sdu */
 	credits = ((chan->_sdu_len - net_buf_frags_len(buf)) +
 		   (chan->rx.mps - 1)) / chan->rx.mps;
-	credits -= k_sem_count_get(&chan->rx.credits);
+	credits -= atomic_get(&chan->rx.credits);
```
Comment (joerchan) [inline subsys/bluetooth/host/l2cap.c]: Unrelated to this PR, but credits is s16_t above? Shouldn't it be u16_t @Vudentz ?

**Review by joerchan — COMMENTED** [review]
So this will replace #21072 then?

File: subsys/bluetooth/host/l2cap.c (hunk 3 above)
Comment (Vudentz) [reply subsys/bluetooth/host/l2cap.c]: We could indeed do that, or a get follow by atomic_dec but then it is not an atomic operation but 2 so there is a window where it could be preempted.

File: subsys/bluetooth/host/l2cap.c (hunk 4 above)
Comment (Vudentz) [reply subsys/bluetooth/host/l2cap.c]: I will check about it, I recall it being change to s16_t for some reason but maybe Im mistaken.

File: subsys/bluetooth/host/l2cap.c (hunk 3 above)
Comment (joerchan) [reply subsys/bluetooth/host/l2cap.c]: The window should be handled by having the atomic_cas in a loop until it succeeds AFAIK.

Comment (Vudentz) [reply subsys/bluetooth/host/l2cap.c]: I guess we don't want a busy loop though otherwise it is not that different than using a spin_lock.

Comment (andrewboie) [reply subsys/bluetooth/host/l2cap.c]: You need to delete this function and do something else, working with the existing atomic operations in Zephyr.

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
NACK on implementing non-standard atomic APIs using spinlocks

**Comment by Vudentz** [comment]
> NACK on implementing non-standard atomic APIs using spinlocks

Done.

File: subsys/bluetooth/host/l2cap.c [hunk 5]
```diff
@@ -1256,6 +1252,17 @@ static void l2cap_chan_seg_sent(struct bt_conn *conn, void *user_data)
 	l2cap_chan_tx_resume(BT_L2CAP_LE_CHAN(chan));
 }
 
+static bool test_and_dec(atomic_t *target)
+{
+	if (!atomic_get(target)) {
+		return false;
+	}
+
```
Comment (andrewboie) [inline subsys/bluetooth/host/l2cap.c]: I think we have a race condition here in between the check to see if target is nonzero, and decrementing the target, you could get preempted in between, or it gets updated on another CPU at the same time.

I recommend you look at lib/os/sem.c, note the use of atomic_cas() with while loops.

File: subsys/bluetooth/host/l2cap.c (hunk 4 above)
Comment (mped-oticon) [reply subsys/bluetooth/host/l2cap.c]: I hope it is u16, since I witness some Androids granting 0xffff credits - and they don't mean -1 credits :)

File: subsys/bluetooth/host/l2cap.c (hunk 5 above)
Comment (Vudentz) [reply subsys/bluetooth/host/l2cap.c]: Interesting if such construction is recommended why we don't we promote them to atomic_bound_dec and atomic_bound_inc? They could be used as is in this case.

Comment (andyross) [reply subsys/bluetooth/host/l2cap.c]: (Oops, fat-fingered the resolve button, sorry).

To be clear: that construction is required.  Consider the case where you enter this function, decide that the count is non-zero, and then between the if() block and the atomic_dec() another thread runs and decrements the count to zero.  This line will then wrap it around to 0xffffffff incorrectly.   Two atomic operations used sequentially is not atomic.

I don't think having a bounded add/dec operation in the kernel would be a bad idea at all, though it's a little out of scope of the existing atomic.h API which is really there just to expose common/portable hardware instructions for atomic operations.

**Review by andyross — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/l2cap.c (hunk 5 above)
Comment (andrewboie) [reply subsys/bluetooth/host/l2cap.c]: the bounded_inc/bounded_dec functions have some special semantics related to how it is used with futexes (bounded dec will decrement to minimum -1 and not minimum).

So I'd be very careful in making sure that this works precisely the way you want if you use that code unmodified.

However, looping with atomic_cas() is a general design pattern to make sure that the value wasn't changed underneath you while you were working on it, so whatever you come up with I think would surely look similar.

Comment (Vudentz) [reply subsys/bluetooth/host/l2cap.c]: Yep, Ive added a pretty similar loop, it just doesn't take a minimum value since all we care is that it doesn't decrement bellow zero.

**Comment by Vudentz** [comment]
@andyross it been taken care now, let me know if there are any other concerns.

File: subsys/bluetooth/host/l2cap.c [hunk 6]
```diff
@@ -1256,6 +1252,22 @@ static void l2cap_chan_seg_sent(struct bt_conn *conn, void *user_data)
...
+		if (!old_value) {
+			return false;
+		}
+
+		new_value = old_value - 1;
+	} while (atomic_cas(target, old_value, new_value) == 0);
+
+	return true;
+}
+
```
Comment (Vudentz) [inline subsys/bluetooth/host/l2cap.c]: @andrewboie @andrewboie this would be the equivalent to i.e. atomic_bound_dec(target, 1) which afaik we could inline in atomic.h, but Im fine either way.

File: subsys/bluetooth/host/l2cap.c (hunk 4 above)
Comment (Vudentz) [reply subsys/bluetooth/host/l2cap.c]: This code is only used to restore RX credits, so it is not influenced by the remote, it restore the amount based on the SDU and then subtract to the existing credits to know if more credits would be needed to complete the SDU, the result can be negative which is why we are using s16_t, we could use a s32_t though that could overflow so it needs to do a bound check as well.

**Review by tyrantlizard77 — APPROVED** [review]
LGTM thanks!

File: subsys/bluetooth/host/l2cap.c (hunk 4 above)
Comment (Vudentz) [reply subsys/bluetooth/host/l2cap.c]: In the end I just did the bound check separately and used u16_t as it should be.

File: subsys/bluetooth/host/l2cap.c [hunk 7]
```diff
@@ -1226,7 +1222,7 @@ static struct net_buf *l2cap_chan_create_seg(struct bt_l2cap_le_chan *ch,
 
 static void l2cap_chan_tx_resume(struct bt_l2cap_le_chan *ch)
 {
-	if (!k_sem_count_get(&ch->tx.credits) ||
+	if (!atomic_get(&ch->tx.credits) ||
 	    (k_fifo_is_empty(&ch->tx_queue) && !ch->tx_buf)) {
 		return;
 	}
```
Comment (andrewboie) [inline subsys/bluetooth/host/l2cap.c]: one question: any complications here if credits gets set to 0 by another CPU in between the atomic_get() and the k_work_submit() call?

Comment (Vudentz) [reply subsys/bluetooth/host/l2cap.c]: There shouldn't be any problem since the code checks again if there any credits on the work itself, that is actually where test_and_dec is used so it would basically just defered itself again until there are credits available.

**Review by andyross — APPROVED** [review]
This looks correct, thanks.

File: subsys/bluetooth/host/l2cap.c [hunk 8]
```diff
@@ -1256,6 +1252,22 @@ static void l2cap_chan_seg_sent(struct bt_conn *conn, void *user_data)
...
+	atomic_t old_value, new_value;
+
+	do {
+		old_value = atomic_get(target);
+		if (!old_value) {
+			return false;
+		}
+
+		new_value = old_value - 1;
+	} while (atomic_cas(target, old_value, new_value) == 0);
```
Comment (joerchan) [inline subsys/bluetooth/host/l2cap.c]: Minor: I would think while(!atomic_cas()); but I see that atomic_cas has different return values based on ifdefs, and that both versions are used elsewhere in the code.

**Review by joerchan — APPROVED** [review]
This was a very nice change.

File: subsys/bluetooth/host/l2cap.c (hunk 7 above)
Comment (andrewboie) [reply subsys/bluetooth/host/l2cap.c]: sounds good

File: subsys/bluetooth/host/l2cap.c (hunk 8 above)
Comment (andrewboie) [reply subsys/bluetooth/host/l2cap.c]: `while(!atomic_cas())` and `while (atomic_cas() == 0)` are the same. MISRA-C wants us to not treat integers as booleans, so the specific == 0 is added.

File: subsys/bluetooth/host/l2cap.c [hunk 9]
```diff
@@ -797,9 +795,7 @@ static void l2cap_chan_rx_give_credits(struct bt_l2cap_le_chan *chan,
 {
 	BT_DBG("chan %p credits %u", chan, credits);
 
-	while (credits--) {
-		k_sem_give(&chan->rx.credits);
-	}
+	atomic_sub(&chan->rx.credits, credits);
```
Comment (jhedberg) [inline subsys/bluetooth/host/l2cap.c]: Why are you doing `atomic_sub()` here and `atomic_add()` in `l2cap_chan_tx_give_credits()`. Shouldn't they both be `atomic_add()`?

**Review by jhedberg — CHANGES_REQUESTED** [review]
See my inline comment about `atomic_sub()` vs `atomic_add()`

File: subsys/bluetooth/host/l2cap.c (hunk 9 above)
Comment (Vudentz) [reply subsys/bluetooth/host/l2cap.c]: Will fix that, not sure how this worked for me but it obviously wrong.

**Review by jhedberg — APPROVED** [review]
